sensor: add driver and example for HX711

GeorgeGkinis opened the pull request:
Signed-off-by: George Gkinis <gkinis.georgios@gmail.com>

carlescufi commented:
@GeorgeGkinis thanks for the contribution. There are 2 commits in this PR with the same commit message. Please squash them together or update the commit messages so they reflect what they do respectively and then force-push to your branch.

galak reviewed (changes requested):
Please convert driver to use new DT macro's.  Look at any other sensor driver to see examples.

GeorgeGkinis commented:
@galak I think it should be ok now.
@carlescufi squashed

GeorgeGkinis commented:
@galak @carlescufi 
Are there any more actions that i should take to proceed with merge?

carlescufi commented:
> @galak @carlescufi
> Are there any more actions that i should take to proceed with merge?

Sorry for the delay, I've asked @MaureenHelm, who is the sensor subsystem maintainer, to review.

MaureenHelm commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+	ret = hx711_wait_until_ready(data, 100);
+
+
+	/* Clock data out. Optionally disable interrupts */
```
why optionally disable interrupts?

GeorgeGkinis replied:
If the **SCK** pin stays **HIGH** for longer than **50 µs** then the **HX711** switches off.

I introduced **CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING** to let the user decide if measuring weight is more important than other aspects of their software that might be IRQ based.

I don't really expect issues but to make sure I disable IRQ for the solution I am working on.

Someone else might not want this, so he can let IRQ's on.
(default **DISABLE_INTERRUPTS_WHILE_POLLING** is **n**)

Leaving as is for now.

MaureenHelm commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+		if (hx711_attr_set_rate(data, val) == 0) {
+			LOG_DBG("Attribute RATE set to %d\n", data->rate);
+		}
+		return ret;
```
Shouldn't you update ret based on the result of `hx711_attr_set_rate`?

GeorgeGkinis replied:
Yes I should.
Done.

MaureenHelm commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+
+	/* Get a reading to set GAIN */
+	ret = hx711_sample_fetch(dev, SENSOR_CHAN_WEIGHT);
+	if (ret != 0) {
+		return ret;
+	}
```
You can remove this block since you always return `ret` below

GeorgeGkinis replied:
Done.

MaureenHelm commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+		avg += data->reading;
+	}
+	LOG_DBG("Average before division : %d", avg);
+	avg = avg / readings;
```
Need to check that `readings != 0`

GeorgeGkinis replied:
Done.

MaureenHelm commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+
+
+DEVICE_AND_API_INIT(hx711_dev, DT_INST_LABEL(0), &hx711_init,
+&hx711_data, &hx711_config,
+POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY, &hx711_api);
```
Please indent these lines

GeorgeGkinis replied:
Done.

MaureenHelm commented on drivers/sensor/sensor_shell.c:
```diff
@@ -52,6 +52,7 @@ const char *sensor_channel_name[SENSOR_CHAN_ALL] = {
 	[SENSOR_CHAN_VOLTAGE] =		"voltage",
 	[SENSOR_CHAN_CURRENT] =		"current",
 	[SENSOR_CHAN_ROTATION] =	"rotation",
+	[SENSOR_CHAN_WEIGHT] =      "weight"
```
Please replace the spaces with tab(s)

GeorgeGkinis replied:
Done.

MaureenHelm commented on dts/bindings/sensor/avia,hx711.yaml:
```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: Apache-2.0
```
Please add a copyright too

GeorgeGkinis replied:
Done.

MaureenHelm commented on samples/sensor/hx711/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
Please move the sample to a separate commit. You can add the commit to this PR, don't need to create a new PR.

GeorgeGkinis replied:
Done.

MaureenHelm replied:
Thanks

MaureenHelm commented on include/drivers/sensor.h:
```diff
@@ -161,6 +161,8 @@ enum sensor_channel {
 	/** State of health measurement in % **/
 	SENSOR_CHAN_GAUGE_STATE_OF_HEALTH,
+	/** Weight. Units depend on SENSOR_ATTR_SLOPE value. */
+	SENSOR_CHAN_WEIGHT,
```
I'd like to see a standard unit here, not variable

GeorgeGkinis replied:
I would really like to be able to force grams as per SI but I believe that it is not possible.
You see the HX711 is simply an ADC, what I get from the sensor is an 24 bit value.
The sensor it self is agnostic to the units, unlike gyroscopes or humidity sensors.

The actual raw sensor value depends on the mechanical and material aspects of the weight scale and also the specific load cells used.

To get an actual measurement which makes any sense in any unit the following formula is used :

**_weight = slope * (raw - offset)_**

**offset** is an raw measurement used for zeroing obtained by **hx711_tare**.
**slope**, well it is the slope of the linear change and is obtained by **hx711_calibrate**.

**hx711_calibrate** requires a target parameter as an **u32_t**.

After zeroing, the user invokes this function and provides a known weight in the unit of his choosing.

If the **_target_** parameter represents _grams_, then **val->val1** from **hx711_channel_get** represents grams.
If it represents _pounds_, then **val->val1** represents pounds.

If I could force **u32_t** to represent **grams** then is could force **SENSOR_CHAN_WEIGHT** to be **grams**.
But as it is, I really cannot force the user to represent grams for the target parameter.

Of course I could **typedef u32_t grams **, but this still would not restrict people from using imperial, tons, Kg or anything they would like.

(**offset** and **tare** can also be defined through **Kconfig** in the case of a controlled manufacturing process of the weight scale.)

MaureenHelm replied:
I'm not sure what alternatives to suggest here, but a variable unit just doesn't make sense with the sensor API. Could you use device tree properties, and set them at the board or application level?

@microbuilder @pabigot any ideas?

microbuilder replied:
@MaureenHelm I agree that we should have specific SI units and scale associated with channels, grams making the most sense here.

Perhaps we have to break this SI unit expectation for certain edge cases, and clearly document that in the sensor driver, but couldn't we use a device tree property to define a `whatever` to `g` coefficient with this sensor, that the firmware author should know for their own hardware, and we can apply that to return `g`?

We really need to come back to the sensor API sometime in the near future, though. I had similar issues designing a spectrometer recently, including a Zephyr driver for the spectral sensor, and there is no good way to handle that data today, similar to this example.

Some units in the sensor API today are just outright wrong as well, such as for light:

```
/** Illuminance in infra-red spectrum, in lux. */
SENSOR_CHAN_IR,
/** Illuminance in red spectrum, in lux. */
SENSOR_CHAN_RED,
/** Illuminance in green spectrum, in lux. */
SENSOR_CHAN_GREEN,
/** Illuminance in blue spectrum, in lux. */
SENSOR_CHAN_BLUE,
```

Lux, as a photometric value, is only relevant to human vision and visible light, and is based on a model that estimates how the human eye perceives brightness with the red, green and blue cones in our eyes. As such, lux has no meaning for individual color channels, and certainly not for IR which it outside the human visible range, and these should all be radiometric units like irradiance.

pabigot replied:
From what I can see from the datasheet and SparkFun tutorials it should work to specify that the unit for `SENSOR_CHAN_WEIGHT` is grams.  Then `avia_hx711_calibrate()` should be given the calibration weight as a `struct sensor_value` in grams.  This can be documented in the (to be provided) documentation for these functions.

A user may choose to provide a calibration value in an arbitrary unit and everything will work, with the understanding that doing so is not compliant with Zephyr's sensor framework.

GeorgeGkinis replied:
There seems no way to force the API to work with **only** grams. 
I agree with @pabigot  that documenting that the values passed to calibrate and the value returned from **sensor_channel_get()** are in grams.

Documentation would then be the only **_line of defence_** against someone using another unit.

I would be a shame though to not let people know they can use **any unit** they like right away, and make them write conversion functions :)

Currently I document that the unit returned is the same as the unit provided during calibration:

```
/**
 * @brief Callibrate the HX711.
 *
 * Given a target value of aknown weight the slope gets calculated.
 * This is actually unit agnostic.
 * If the target weight is given in grams, lb, Kg or any other weight unit,
 * the slope will be calculated accordingly.
 *
 * @param dev Pointer to the hx711 device structure
 * @param target Target weight in whatever unit.
 * @param readings Number of readings to take for calibration.
 *        5~10 readings should be enough, although more are allowed.
 * @retval The slope value
 *
 */
```

GeorgeGkinis replied:
I changed documentation indicating that WEIGHT is in grams.
The possibility to use other units is described.
```
/**
 * @brief Callibrate the HX711.
 *
 * Given a target value of aknown weight the slope gets calculated.
 * This is actually unit agnostic.
 * If the target weight is given in grams, lb, Kg or any other weight unit,
 * the slope will be calculated accordingly.
 *
 * @param dev Pointer to the hx711 device structure
 * @param target Target weight in grams.
 *        If target is represented in another unit (lb, oz, Kg) then the
 *        value returned by sensor_channel_get() will represent that unit.
 * @param readings Number of readings to take for calibration.
 *        5~10 readings should be enough, although more are allowed.
 * @retval The slope value
 *
 */
```

I believe that with this change the solution can be accepted.

MaureenHelm commented on samples/sensor/hx711/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+    harness: console
+    tags: sensors
+    depends_on: gpio
+    filter: dt_compat_enabled("avia,hx711")
```
I don't usually like whitelists, but in this case you need it because there aren't any boards with this compatible. Please replace with `platform_whitelist: nrf52_pca10040`.

That said, didn't this board get renamed recently? @ioannisg ?

ioannisg replied:
Yeap, that's now called nrf52dk_nrf52832

GeorgeGkinis replied:
I do not really understand what **sample.yml** does.

Do I have to change the contents to 
```
sample:
  name: HX711 24-Bit Analog-to-Digital Converter (ADC) for Weigh Scales
tests:
  sample.sensor.hx711:
    platform_whitelist: nrf52dk_nrf52832
```
?

MaureenHelm replied:
`sample.yaml` helps sanitycheck determine which sample applications should be built and tested on which boards.
https://docs.zephyrproject.org/latest/guides/test/sanitycheck.html

GeorgeGkinis replied:
contents changed to :

```
sample:
  name: HX711 24-Bit Analog-to-Digital Converter (ADC) for Weigh Scales
tests:
  sample.sensor.hx711:
    tags: sensors
    depends_on: gpio
    platform_whitelist: nrf52dk_nrf52832

```

GeorgeGkinis replied:
Is this change acceptable?
Can this also be resolved now?

MaureenHelm reviewed (changes requested):
Please add this sensor to `tests/drivers/build_all/sensors_a_h.conf`

Thanks for the contribution!

GeorgeGkinis commented:
@MaureenHelm 
Thanks for your review!

Can I fix every issue with a separate commit, or should I do 
```git rebase -i e0b47aa7ed3c41cba7930ebfd95b7171b21ef9f2```

as described in [contribution workflow](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow)
step 15 onwards.

The easiest way for me is just another git commit on top of the latest, but I am not sure if it will break things.

Please keep in mind that this is my first **pull request/contribution _ever_** and is an intense and rewarding learning experience for me.

GeorgeGkinis commented:
Sorry for all the strange force-pushes.
I almost lost my work, luckily I had another checkout locally.

Resuming with fixing the review items.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+static int hx711_wait_until_ready(struct hx711_data *data, uint32_t usec)
+{
+	while (!hx711_is_ready(data) && usec >= 0) {
+		k_busy_wait(1);
```
You're definitely right.
Done.

GeorgeGkinis replied:
I tried **k_sleep(K_USEC(1));** but then the thread seems to become unresponsive and slow.
No clue why this is. I suspect I have to manually wake up the thread somehow.

I changed it back to **k_busy_wait(1);** and the code flies.

As mentioned [here ](https://docs.zephyrproject.org/1.9.0/kernel/threads/scheduling.html#busy-waiting) if the wait period is small there is sometimes no justification for context switch to take place.

**k_sleep** could be used in conjunction with the **falling edge** of the **DOUT** pin to wake the thread up through an **GPIO interrupt trigger**.
It would be tidy and snappy, but for now It seems OK.

Unless the **1usec** delay seems too big to busy wait I would rather start using the library for my application.

GeorgeGkinis replied:
I changed this bit to use k_sleep(K_USEC(10)) as you suggested.

The time needed for HX711 to power up is 566ms. I use usec == 600000 now.
It is also the maximum amount of time observed for the HX711 to be ready to provide  a reading.

If that time is elapsed -EACCES is returned.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+ * @retval 0 When data is available,
+ * @retval -EACCES error if usec's are elapsed with no data available.
+ */
+static int hx711_wait_until_ready(struct hx711_data *data, uint32_t usec)
```
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,528 @@
+
+#include "hx711.h"
+
+LOG_MODULE_REGISTER(HX711, CONFIG_LOG_DEFAULT_LEVEL);
```
Done.

GeorgeGkinis commented on drivers/sensor/hx711/Kconfig:
```diff
@@ -0,0 +1,82 @@
+
+config HX711_DISABLE_INTERRUPTS_WHILE_POLLING
+	bool "Disable interrupts while polling"
+	default n
```
Done.

GeorgeGkinis commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -26,6 +26,7 @@ add_subdirectory_ifdef(CONFIG_TI_HDC		ti_hdc)
 add_subdirectory_ifdef(CONFIG_HMC5883L		hmc5883l)
 add_subdirectory_ifdef(CONFIG_HP206C		hp206c)
 add_subdirectory_ifdef(CONFIG_HTS221		hts221)
+add_subdirectory_ifdef(CONFIG_HX711         hx711)
```
Done.

MaureenHelm commented:
> Can I fix every issue with a separate commit, or should I do
> `git rebase -i e0b47aa7ed3c41cba7930ebfd95b7171b21ef9f2`
> 
> as described in [contribution workflow](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow)
> step 15 onwards.

Please do the latter

> Please keep in mind that this is my first **pull request/contribution _ever_** and is an intense and rewarding learning experience for me.

No worries, thanks for contributing!

microbuilder commented:
> Sorry for all the strange force-pushes.
> I almost lost my work, luckily I had another checkout locally.
> 
> Resuming with fixing the review items.

No worries, it's a complicated process the first couple times you do it, and git can at times be obtuse or opaque since there are often many ways to do similar things. Keep notes as you go for future you, and doing is the best way to learn.

pabigot commented:
> The easiest way for me is just another git commit on top of the latest, but I am not sure if it will break things.

I personally always do incremental commits, so if something goes wrong I can go back.  It also allows me to check in changes that go with a specific commit (e.g. for the driver, or for the sample).  But before you push your accumulated changes for the next review you should do an interactive rebase to fixup to the minimal set of changes.

pabigot reviewed (commented):
The advice on how to add driver-specific API for a subsystem like sensors has changed.  The documentation was partially updated in #24567, and #24741 provides more complete information.

The tare and calibrate functions should be configured to use the new recommended interface: standalone functions, not part of an extended API structure.  Following the initial example of using this extension mechanism (in #17531) I would prefer to see the headers and specific function prefixes all use `avia_hx711`, to match the compatible.  "hx711" is simply too generic IMO.  It may be that it should be `sensor_avia_hx711`, but that's something we'll need to discuss in the API telecon.

Thanks for your contribution.  This driver has features that put it on the edge of what the sensor API (and driver API in general) was designed to handle, so it may take a while before we all agree on the best way to express things.

GeorgeGkinis commented:
> The tare and calibrate functions should be configured to use the new recommended interface: standalone functions, not part of an extended API structure. Following the initial example of using this extension mechanism (in #17531) I would prefer to see the headers and specific function prefixes all use `avia_hx711`, to match the compatible. "hx711" is simply too generic IMO. It may be that it should be `sensor_avia_hx711`, but that's something we'll need to discuss in the API telecon.

I personally kind of liked the extended api methodology. If you check the sample application I included **<drivers/sensor/hx711.h>** where the **hx711_driver_api** struct gets exposed.
Then I do a typecast and can now use the extended api. I can even name the api as I prefer.

It may of course be the case that I am doing things wrong security-wise. Concepts like user-space and kernel-space are not really clear to me as yet.

Would the team advise me to wait for an **updated driver API** of version **2.3**? 
I am kinda short on time between my job and kids. This is just a hobby :)

pabigot commented:
> I personally kind of liked the extended api methodology. If you check the sample application I included **<drivers/sensor/hx711.h>** where the **hx711_driver_api** struct gets exposed.
> Then I do a typecast and can now use the extended api. I can even name the api as I prefer.

I liked it too, which is why I proposed it last year.  The problem is that drivers that use this technique can never support userspace systemcalls, because they way those are handled involves a post-link step where the relevant subsystem (in this case `sensor`) is identified by the DWARF type information associated with the API structure object.  In this case the API pointer is to an object of type `struct hx711_driver_api hx711_api` which is not a registered driver subsystem, so if `CONFIG_USERSPACE=y` the system calls for the entire application will incorrectly generated.

You don't need to wait for anything; just use `struct sensor_driver_api hx711_api` and make the extended methods public.  The in-flight PRs are just to correct the recommendations to be a solution that works.  The CCS811 sensor driver in-tree does exactly this for its `ccs811_baseline_fetch/update` extended methods.  They don't currently support userspace calls, so you can use the same technique; in both cases the extensions can be converted to userspace-capable without changing the API.

For information on userspace see [this documentation](https://docs.zephyrproject.org/latest/reference/usermode/index.html); the Constraints section near the bottom has information on system call support.  I believe it's not often used, but it's very important to some people.

GeorgeGkinis commented:
> The advice on how to add driver-specific API for a subsystem like sensors has changed. The documentation was partially updated in #24567, and #24741 provides more complete information.
> 
> The tare and calibrate functions should be configured to use the new recommended interface: standalone functions, not part of an extended API structure. Following the initial example of using this extension mechanism (in #17531) I would prefer to see the headers and specific function prefixes all use `avia_hx711`, to match the compatible. "hx711" is simply too generic IMO. It may be that it should be `sensor_avia_hx711`, but that's something we'll need to discuss in the API telecon.
> 
> Thanks for your contribution. This driver has features that put it on the edge of what the sensor API (and driver API in general) was designed to handle, so it may take a while before we all agree on the best way to express things.

I moved the sensor specific functions outside of the sensor_driver_api.
They reside in sensor/hx711.h.

microbuilder commented on samples/sensor/hx711/prj.conf:
```diff
@@ -0,0 +1,29 @@
+CONFIG_LOG_PROCESS_THREAD_STACK_SIZE=1024
+
+CONFIG_CONSOLE=y
+CONFIG_USE_SEGGER_RTT=y
```
The `RTT` entries might be better in a board-specific conf file in the boards folder, since they won't be available or meaningful defaults on every board, and only items common to every board should be placed in this file.

MaureenHelm replied:
Agreed. I think most of the LOG configurations should also be removed.

GeorgeGkinis replied:
They have been removed.
I do not know why they still show up here....

microbuilder commented on include/drivers/sensor/hx711.h:
```diff
@@ -0,0 +1,64 @@
+/**
+ * @brief Callibrate the HX711.
+ *
+ * Given a target value of aknown weight the slope gets calculated.
```
Minor, but small typo with `aknown`.

GeorgeGkinis replied:
Fixed.

microbuilder commented on samples/sensor/hx711/src/main.c:
```diff
@@ -0,0 +1,51 @@
+		LOG_INF(" %d..", i);
+		k_msleep(1000);
+	}
+
```
Minor, but extra line should be removed.

microbuilder commented on samples/sensor/hx711/prj.conf:
```diff
@@ -0,0 +1,29 @@
+CONFIG_HX711=y
+CONFIG_GPIO=y
+CONFIG_HX711_CALIBRATION_WEIGHT=500
+
```
Please remove extra space.

GeorgeGkinis replied:
Done

microbuilder commented:
@GeorgeGkinis Just as a helpful FYI, you might want to look at https://github.com/zephyrproject-rtos/zephyr/blob/master/.uncrustify.cfg, which you can use with the `uncrustify` tool to automatically format your code to use the Zephyr coding style. See: https://docs.zephyrproject.org/latest/contribute/index.html#uncrustify

I used a package in Atom called *Atom Beautify* myself, pointing to this file for C code, and then can just do a CTRL+OPT+B to auto cleanup when I've made some changes. I'm sure there are options for other editors, whatever you prefer to use yourself.

microbuilder commented:
@GeorgeGkinis I ordered two versions of this sensor to be able to test this. Let me know when you're happy with the driver, and I can solder one together and test.

GeorgeGkinis commented:
Wow!
What happened with all those review requests?   
I rebased my branch on top ofmaster and made some changes. 
Then pushed to origin (my own fork). 
Is there an automatic review request introduced to the build system?

pabigot commented:
You didn't rebase, you created a merge commit between master and your branch.  The result is that every commit added to master since the base commit 2d9baf97f9 of your branch now appears to be in your PR.  Try the rebase again.

GeorgeGkinis commented:
Sorry for the mess @pabigot :( 
I Checked my history : 

 git fetch --all
 git rebase --ignore-whitespace upstream/master
 git stash
 **git pull <- Damn!**
 git push --force origin

I just did : 
```
git reflog show
q@PPWOR239:/mnt/c/FUN/NRF5/hx711/zephyr$ git reflog show
b6490c8ebb (HEAD -> Add_HX711_weight_scale, origin/Add_HX711_weight_scale) HEAD@{0}: reset: moving to HEAD@{6}
e163db6b41 HEAD@{1}: reset: moving to HEAD
e163db6b41 HEAD@{2}: commit: Added RATE change to example
8895692088 HEAD@{3}: commit: Fixed typo
541186e7e0 HEAD@{4}: commit: Removed empty lines
ad23cb7907 HEAD@{5}: commit: Added rate pin configuration in devicetree
c496c223cb HEAD@{6}: commit: Removed unnesecary KConfig options
b6490c8ebb (HEAD -> Add_HX711_weight_scale, origin/Add_HX711_weight_scale) HEAD@{7}: pull: Merge made by the 'recursive' strategy.
a9a06e819d HEAD@{8}: reset: moving to HEAD
a9a06e819d HEAD@{9}: rebase finished: returning to refs/heads/Add_HX711_weight_scale
a9a06e819d HEAD@{10}: rebase: sensor: add driver for HX711
045112c7b8 HEAD@{11}: rebase: sensor: add example for HX711
1a89ca1238 (upstream/master) HEAD@{12}: rebase: checkout upstream/master
300b90f340 HEAD@{13}: reset: moving to HEAD
300b90f340 HEAD@{14}: checkout: moving from master to Add_HX711_weight_scale
47f1665e78 (origin/master, origin/HEAD, master) HEAD@{15}: clone: from https://github.com/GeorgeGkinis/zephyr.git

git reset --keep HEAD@{9}

git push --force origin Add_HX711_weight_scale
```
Hope it's ok now..

I also made a copy of my changes. 
Gonna put them on top again.

GeorgeGkinis commented:
> @GeorgeGkinis I ordered two versions of this sensor to be able to test this. Let me know when you're happy with the driver, and I can solder one together and test.

@microbuilder I think I am done. I will sleep on it and check my code for the coming days again.

The option to choose sampling rate is implemented and tested along with putting the module to sleep.

I had to desolder the RATE pin on my module though as it was connected to ground.

Please try to run the sample and do a review whenever you got some spare time.

Should I upload a pulseview capture somewhere?
I used sigrok to test with a logic analyser.

Cheers!

microbuilder commented:
@GeorgeGkinis Thanks, I'll hook this up early next week and post a proper review.

microbuilder commented:
@GeorgeGkinis You will have to resolve the issues in the checkpatch comments at the top of this PR, though ("line over 80 characters", etc.), before we can merge, just as a FYI.

MaureenHelm commented on include/drivers/sensor.h:
```diff
@@ -277,6 +280,18 @@ enum sensor_attribute {
+	 * The sensor value returned will be altered by the amount indicated by
+	 * slope: final_value = sensor_value * slope.
+	 */
+	SENSOR_ATTR_SLOPE,
+
+	/**
+	 * The sensor gain.
+	 */
+	SENSOR_ATTR_GAIN,
```
These can now be removed

GeorgeGkinis replied:
I do not understand why they need to be removed.
How are those attributes supposed to be changed during runtime?

**sensor_attr_set** still requires an **enum sensor_attribute attr**

MaureenHelm commented on samples/sensor/hx711/README.rst:
```diff
@@ -10,13 +10,37 @@ Offset is calculated without any weight put on the scale.
 calibration is performed.
 
-After that calibration the weight is printed every second.
+After calibration :
```
all of the sample-related changes should be in commit `sensor: add example for HX711`. Some are incorrectly in commit `sensor: add driver for HX711`

GeorgeGkinis replied:
You are right. 
Fixed.

nobodyguy commented:
How can I help this PR to be merged?

henrikbrixandersen commented:
@GeorgeGkinis Please rebase on master and resolve the merge conflicts.

nobodyguy commented:
@GeorgeGkinis Do you need any help with that?

aFooBar commented:
> @GeorgeGkinis Do you need any help with that?
I managed to get it work by changing TIMER_COUNTERTOP from 32 to 64
Guess there's a problem with my clock definition; did you modify the default clock setup (I'm using the default nRF example config) ?

Many thanks for your driver

GeorgeGkinis commented:
Hello @henrikbrixandersen

I am currently on vacation and could work on resolving the conflicts begin september, as I did not bring the HW with me to test after merge :)

GeorgeGkinis commented:
Hello @henrikbrixandersen,

Merged on top of master and confilcts resolved.

@nobodyguy, @microbuilder, 
Can either (or both) of you verify this also works on your HW?

Thanks.

nobodyguy commented:
I'll check during this week and let you know.

GeorgeGkinis commented:
Hello @henrikbrixandersen,

What is needed to merge?
Since it has been a while I suspect a rebase on top of master again?

@nobodyguy, @microbuilder

Any luck with verifying?

erwango commented on samples/sensor/hx711/boards/nrf52dk_nrf52832.overlay:
```diff
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/{
+hx711 {
```
Please fix indentation on this line

erwango commented on samples/sensor/hx711/boards/nrf52dk_nrf52832.overlay:
```diff
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
Missing copyright

erwango commented on samples/sensor/hx711/README.rst:
```diff
@@ -0,0 +1,99 @@
+Requirements
+************
+
+This sample uses the HX711 sensor module from AliExpress, although any HX711 hardware implementation should work.
```
"HX711 sensor module from AliExpress" Please provide more detailed information (maker reference would be better than the shop adress)

GeorgeGkinis replied:
I agree about not posting an URL for the AliExpress shop, but my module does not really have a model number on it..

Should I just remove the URL?

GeorgeGkinis replied:
Added board model used : "shengyang XFW-HX711"

erwango commented on samples/sensor/hx711/sample.yaml:
```diff
@@ -4,4 +4,4 @@ tests:
     tags: sensors
     depends_on: gpio
-    platform_whitelist: nrf52dk_nrf52832
+    platform_allow: nrf52dk_nrf52832
```
Please fix in initial commit

erwango commented on samples/sensor/hx711/src/main.c:
```diff
@@ -15,7 +15,7 @@
 #include <logging/log.h>
 
-LOG_MODULE_REGISTER(main, LOG_LEVEL_DBG);
+LOG_MODULE_REGISTER(main, LOG_LEVEL_INF);
```
Please fix in initial commit

erwango commented:
> What is needed to merge?

You need 2 approvals.
I've re-requested review to all people involved.
Adding the V2.5.0 milestone tag to be sure it won't be missed

erwango commented:
@GeorgeGkinis do you plan to continue this PR ?

GeorgeGkinis commented:
> @GeorgeGkinis do you plan to continue this PR ?

I hooked everything up again and plan resume this week.

GeorgeGkinis committed:
sensor: add driver for HX711

Based on Avia Semiconductor's patented technology,
HX711 is a precision 24-bit analog-to-digital converter (ADC)
designed for weigh scales and industrial control applications
to interface directly with a bridge sensor.

Signed-off-by: George Gkinis <gkinis.georgios@gmail.com>

GeorgeGkinis committed:
sensor: add example for HX711

Added an example for the HX711 driver as separate commit.

Signed-off-by: George Gkinis <gkinis.georgios@gmail.com>

GeorgeGkinis commented:
Hmm...

Since v2.4.0 "_All device instances got a const qualifier. So this applies to all APIs manipulating struct device * (ADC, GPIO, I2C, …)..."_

GeorgeGkinis commented:
Hello,

I get an error during the compliance tests : 
"**Path '/subsys/usb/class/usb_dfu.c' not found in the tree but is listed in CODEOWNERS**"

It indeed does not exist on the master branch.

nobodyguy commented:
> @nobodyguy, @microbuilder
> 
> Any luck with verifying?

Sorry for my huge delay, I was not able to compile the module out-of-tree with older Zephyr, so I had to postpone the testing until upgrade of my project to the latest version is finished.
Once again thank you a lot for your amazing work and all your precious time.
I did some testing during the last couple of days and it works as expected. Although I noticed that information about fractional part of a reading is lost in `hx711_channel_get` function. See the code comment I added.

I also noticed that there is already a second PR with HX711 support - https://github.com/zephyrproject-rtos/zephyr/pull/30656 :relaxed:
It would be great if we could maybe unify the work and efford together with @coderkalyan and provide single robust driver with all the great features like interrupt driven reading (seen in https://github.com/zephyrproject-rtos/zephyr/pull/30656) - what do you think about it? :muscle:

nobodyguy commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+	switch (chan) {
+	case SENSOR_CHAN_WEIGHT:
+	case SENSOR_CHAN_ALL:
+		val->val1 = sensor_value_to_double(&data->slope)
```
I would expect fractional part of the result to be available in `val2`, so something like this instead:
```c
case SENSOR_CHAN_ALL: {
	double result = sensor_value_to_double(&data->slope) *
			(data->reading - data->offset);
	val->val1 = (int32_t)result;
	val->val2 = (result - (int32_t)result) * 1000000;
	return 0;
}
```

GeorgeGkinis replied:
Absolutely Right.

GeorgeGkinis replied:
@nobodyguy, 
The only reason I opted then to only use **val1** was code simplification.

After thinking about it for awhile I believed it not to be _necessary_, as it would involve providing the **_target_** argument to _**avia_hx711_calibrate**_ as a **float**, which in turn would use a bit more **floating point arithmetic** and **sensor_value_to_double** calls.

Also currently **CONFIG_HX711_CALIBRATION_WEIGHT** is defined in KConfig.
Last time I checked floats were not supported so I decided to not do fractionals in  order to be able to provide configuration without hardcoding.

we could of course do
**CONFIG_HX711_CALIBRATION_WEIGHT_INTEGER_PART** and **CONFIG_HX711_CALIBRATION_WEIGHT_FRACTIONAL_PART**  but seems quite ugly  imho.

In a nutshell : if someone wants to use _fractional_ **Kg**, he should measure in **gr**  and divide :)

It is surely not unfeasible, and there are also the  **hx711_attr_set_slope** and **hx711_attr_set_offset** functions available.

What do you think about this?

GeorgeGkinis replied:
Fixed.

retfie commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+		if (hx711_attr_set_gain(dev, val) == 0) {
+			LOG_DBG("Attribute GAIN set to %d\n", data->gain);
+		}
+		return ret;
```
Please update **ret** after hx711_attr_set_gain(). It invokes hx711_sample_fetch() and it's possible to change it in case of failure.

GeorgeGkinis replied:
Right.

GeorgeGkinis replied:
done.

retfie commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+{
+	int32_t avg = 0;
+	struct hx711_data *data = dev->data;
+
```
Isn't it better to check readings for zero here and exit. There isn't much point to go further in such case.

GeorgeGkinis replied:
The reading is rarely zero. 
something like one in a million :)

GeorgeGkinis replied:
Won't fix.

retfie commented on samples/sensor/hx711/src/main.c:
```diff
@@ -0,0 +1,114 @@
+/*
```
Shouldn't this file be part of example commit ? https://github.com/zephyrproject-rtos/zephyr/pull/24295/commits/4eb64a10e35e7f14fa9b835d9b22fb4d46ede369

GeorgeGkinis replied:
Right.

retfie commented on tests/drivers/build_all/app.overlay:
```diff
@@ -116,6 +116,15 @@
 		};
 
+		hx711 {
+        	compatible = "avia,hx711";
```
Wrong indentation. Mixed tabs + spaces, and probably one more tab is needed.

GeorgeGkinis replied:
Right.

GeorgeGkinis replied:
Fixed.

retfie commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+}
+
+/**
+ * @brief Set HX711 offset.
```
typo in brief: offset -> slope

GeorgeGkinis replied:
Right.

retfie commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+/**
+ * @brief Callibrate the HX711.
+ *
+ * Given a target value of aknown weight the slope gets calculated.
```
typo:  "aknown"

GeorgeGkinis replied:
Right.

retfie commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+	case HX711_RATE_10HZ:
+	case HX711_RATE_80HZ:
+		data->rate = val->val1;
+		ret = gpio_pin_set(data->rate_gpio,
```
Please check if rate_gpio is not null before call gpio_pin_set(), as 'rate_gpios' is optional into DTS. Example crashes on first call of set_rate() if rate gpio not defined into device tree.

GeorgeGkinis replied:
Absolutely right :)

GeorgeGkinis replied:
Resolved by using an added field **cfg->rate_enabled** which is set to 1 if **rate_gpios** can be found in **DT**.

joelrietzle commented:
@GeorgeGkinis @retfie @galak will this be merged to the master to be compatible with the latest RTOS? Would love to be able to implement this for my latest IoT development project for my company.

microbuilder commented:
@GeorgeGkinis The issue I have is that I have this IC here, but I don't have a load cell to test with. Is there a diagnostic test you can run to make sure the basic driver is working fine, without a load cell present? Of perhaps someone else has the driver plus a load cell to verify that things work as expected (sane units returned, etc.)?

microbuilder commented:
@GeorgeGkinis I just ordered a 5 KG load cell off Amazon, and will test this Monday when it arrives.

galak commented:
> @GeorgeGkinis @retfie @galak will this be merged to the master to be compatible with the latest RTOS? Would love to be able to implement this for my latest IoT development project for my company.

@joelrietzle We're in a stabilization phase for the 2.5 release so new code like this must what til after that release completes.  Additionally the code needs to get 2 approvals for this to be merged.

microbuilder commented:
@GeorgeGkinis I setup a load cell, and tried to build this, but get a lot of devicetree macro errors causing the build to fail:

```bash
$ west build -p auto -b lpcxpresso55s69_cpu0 zephyr/samples/sensor/hx711/ --
...
/Users/kevin/Linaro/zephyr/upstream/zephyr/include/devicetree.h:297:31: note: in expansion of macro 'UTIL_CAT'
  297 | #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
      |                               ^~~~~~~~
/Users/kevin/Linaro/zephyr/upstream/zephyr/include/devicetree.h:1589:27: note: in expansion of macro 'DT_INST'
 1589 | #define DT_DRV_INST(inst) DT_INST(inst, DT_DRV_COMPAT)
      |                           ^~~~~~~
/Users/kevin/Linaro/zephyr/upstream/zephyr/include/devicetree.h:1611:42: note: in expansion of macro 'DT_DRV_INST'
 1611 | #define DT_INST_PROP(inst, prop) DT_PROP(DT_DRV_INST(inst), prop)
      |                                          ^~~~~~~~~~~
/Users/kevin/Linaro/zephyr/upstream/zephyr/include/devicetree.h:1658:29: note: in expansion of macro 'DT_INST_PROP'
 1658 | #define DT_INST_LABEL(inst) DT_INST_PROP(inst, label)
      |                             ^~~~~~~~~~~~
/Users/kevin/Linaro/zephyr/upstream/zephyr/drivers/sensor/hx711/hx711.c:560:32: note: in expansion of macro 'DT_INST_LABEL'
  560 | DEVICE_AND_API_INIT(hx711_dev, DT_INST_LABEL(0), &hx711_init,
      |                                ^~~~~~~~~~~~~
```

microbuilder commented:
> @GeorgeGkinis I setup a load cell, and tried to build this, but get a lot of devicetree macro errors causing the build to fail:

Sorry, I missed the way this drives the IC via GPIO pins (assumed it was I2C) and the need for a board-specific DTS overlay:

```
	hx711 {
		compatible = "avia,hx711";
		status = "okay";
		label = "HX711";
		dout-gpios = <&gpio0 30 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP) >;
		sck-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>;
		rate-gpios = <&gpio0 31 GPIO_ACTIVE_HIGH>;
	};
```

I'll update for the board I'm testing on and try again, since it's useful to test on a new board anyway.

joelrietzle commented:
> > @GeorgeGkinis I setup a load cell, and tried to build this, but get a lot of devicetree macro errors causing the build to fail:
> 
> Sorry, I missed the way this drives the IC via GPIO pins (assumed it was I2C) and the need for a board-specific DTS overlay:
> 
> ```
> 	hx711 {
> 		compatible = "avia,hx711";
> 		status = "okay";
> 		label = "HX711";
> 		dout-gpios = <&gpio0 30 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP) >;
> 		sck-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>;
> 		rate-gpios = <&gpio0 31 GPIO_ACTIVE_HIGH>;
> 	};
> ```
> 
> I'll update for the board I'm testing on and try again, since it's useful to test on a new board anyway.

Awesome, I'm excited! I will try out some of the stuff that you've done already next week. :)

GeorgeGkinis commented:
Excuses for the delay everyone, I had some health related adventures.

All is good now, I hope that I can wrap up the review issues shortly.

microbuilder commented on samples/sensor/hx711/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_SENSOR=y
+CONFIG_HX711=y
+CONFIG_GPIO=y
+CONFIG_HX711_CALIBRATION_WEIGHT=500
```
It likely makes sense to enable logging here, since the sample has no real use if you don't enable it.

```suggestion
CONFIG_HX711_CALIBRATION_WEIGHT=500
CONFIG_LOG=y
```

GeorgeGkinis replied:
Done.

microbuilder commented:
I set this up with an LPCXpresso55S69, just because it was on my desk, using the following DTS overlay file:

```dts
/*
 * Copyright (c) 2021 Linaro Ltd.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

/{
	hx711 {
		/* Tested with the Mikroe Load Cell Click board (MIKROE-3168). */
		compatible = "avia,hx711";
		status = "okay";
		label = "HX711";
		dout-gpios = <&gpio1 18 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP) >;
		sck-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>;
		rate-gpios = <&gpio1 1 GPIO_ACTIVE_HIGH>;
	};
};
```

I'll have to find a calibration weight to properly test the sample app, though. I think I have some here SOMEWHERE from a previous life.

GeorgeGkinis commented:
> I set this up with an LPCXpresso55S69, just because it was on my desk, using the following DTS overlay file:
> 
> `boards/lpcxpresso55s69_cpu0.overlay`:
> 
> ```
> /*
>  * Copyright (c) 2021 Linaro Ltd.
>  *
>  * SPDX-License-Identifier: Apache-2.0
>  */
> 
> /{
> 	hx711 {
> 		/* Tested with the Mikroe Load Cell Click board (MIKROE-3168). */
> 		compatible = "avia,hx711";
> 		status = "okay";
> 		label = "HX711";
> 		dout-gpios = <&gpio1 18 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP) >;
> 		sck-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>;
> 		rate-gpios = <&gpio1 1 GPIO_ACTIVE_HIGH>;
> 	};
> };
> ```
> 
> I'll have to find a calibration weight to properly test the sample app, though. I think I have some here SOMEWHERE from a previous life.

I use another **prj.conf** to run the sample.
CONFIG_HX711_CALIBRATION_WEIGHT=491 which is the weight of a glass ball I weighted with my kitchen scale :)

pabigot commented:
> I'll have to find a calibration weight to properly test the sample app, though. I think I have some here SOMEWHERE from a previous life.

A [US quarter dollar weighs 0.200 ounces](https://cointrackers.com/blog/44/how-much-does-a-quarter-weigh/) and a nickel weighs 2.5 g, so these are useful for calibration.

GeorgeGkinis commented:
Also please mind to use an apropriate weight for the load cell you purchased.

Do not use a 10gr weight for a 5Kg load cell :)

microbuilder commented:
> > I'll have to find a calibration weight to properly test the sample app, though. I think I have some here SOMEWHERE from a previous life.
> 
> A [US quarter dollar weighs 0.200 ounces](https://cointrackers.com/blog/44/how-much-does-a-quarter-weigh/) and a nickel weighs 2.5 g, so these are useful for calibration. If you're in the US. Presumably other coinage has similar standard weights.

That's good to know, thanks. I did find a standard 100g calibration weight, though. If that isn't enough for the 5 KG load cell, I'll just measure something bigger on a high-quality scale as a reference.

microbuilder commented:
@GeorgeGkinis There doesn't seem to be any sort of 'sanity check' at startup, which is always a good idea in the init function. Usually, I'll just read something like the chip ID or a 'whoami' or known-value register at startup to make sure it's the expected value, and we can continue with confidence that the device is properly connected. This chip doesn't seem to have 'registers', though. Any thoughts on how we can perform a simple sanity check at startup to verify wiring, etc.?

GeorgeGkinis commented:
> @GeorgeGkinis There doesn't seem to be any sort of 'sanity check' at startup, which is always a good idea in the init function. Usually, I'll just read something like the chip ID or a 'whoami' or known-value register at startup to make sure it's the expected value, and we can continue with confidence that the device is properly connected. This chip doesn't seem to have 'registers', though. Any thoughts on how we can perform a simple sanity check at startup to verify wiring, etc.?

Actually I have seen in practice (and lost quite some time) because of bad wiring, namely kernel panics and freezes :)

During the **_hx711_init_** function **_hx711_sample_fetch_** is called which in turn invokes **_hx711_wait_until_ready_** which returns **_-EACCES_** if more than **_600000usec_** _(hardcoded)_ have passed, which is bubbled up as return from **_hx711_init_**

At **_DEVICE_AND_API_INIT_** the priority is configured as **_CONFIG_SENSOR_INIT_PRIORITY_**

Would I be out of line expecting **-EACCES** to be handled by the kernel gracefully?
Should I use another priority or return another error during from **_hx711_init_**?

pabigot commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,562 @@
+	LOG_DBG("Set DOUT pin : %d\n", cfg->dout_pin);
+
+	/* Get a reading to set GAIN */
+	ret = hx711_sample_fetch(dev, SENSOR_CHAN_WEIGHT);
```
Referencing https://github.com/zephyrproject-rtos/zephyr/pull/24295#issuecomment-779216345: This is problematic.

The init function is invoked by the kernel in a loop along with all other drivers at the same init level (`POST_KERNEL` in this case.)  Having this function stall for up to 60 s is not acceptable, as it would prevent startup from initializing other drivers and getting to the application.

joelrietzle replied:
What is the progress on this right now?

GeorgeGkinis replied:
Unfortunately, after a powerdown or power reset the HX711 needs approximately 600ms to provide the first measurement (dout from high to low).

I could avoid calling **hx711_sample_fetch** during init but then there would not be any guarantee that the hardware is functioning properly.

Also I still would like to know what to do in case of bad wiring.
As is said as it is right now I get kernel panics during init if the module is not connected.

Not sure if this happens because 600ms is too much or because **-EACCESS** is returned from init.

Should we allow init to complete and get **-EACCESS** during the first call to **hx711_sample_fetch**?

GeorgeGkinis replied:
Never got an answer to this one. With the help of @nobodyguy we now use semaphores and interrupts, but the 600ms will still be there as it is hardware specific to HX711 as explained.

hx711_sample_fetch during init should not be avoided as it sets the appropriate GAIN and RATE according to KConfig.

I still get kernel panics if the sensor is not connected (bad wiring). This happens if **hx711_sample_fetch** returns an error (-EIO). 

I added an appropriate remark "Is the device properly connected?" but do not know how to circumvent this issue.

```
[00:00:01.034,332] <err> HX711: Weight data not ready within 700 ms. Is the device properly connected?
*** Booting Zephyr OS build v3.0.0-rc1-307-ga775606ae955  ***
[00:00:01.034,423] <inf> main: device is (nil), name is  /
[00:00:01.034,423] <inf> main: Calculating offset...
[00:00:01.034,454] <err> os: ***** MPU FAULT *****
[00:00:01.034,454] <err> os:   Data Access Violation
[00:00:01.034,454] <err> os:   MMFAR Address: 0x34a5
[00:00:01.034,484] <err> os: r0/a1:  0x04e000e1  r1/a2:  0x00000039  r2/a3:  0x00000003
[00:00:01.034,484] <err> os: r3/a4:  0x04e000e1 r12/ip:  0x00000000 r14/lr:  0x0000735d
[00:00:01.034,515] <err> os:  xpsr:  0x61000000
[00:00:01.034,515] <err> os: Faulting instruction address (r15/pc): 0x0000734c
[00:00:01.034,545] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:01.034,576] <err> os: Current thread: 0x200012b8 (main)
[00:00:09.903,228] <err> os: Halting system

```

Are errors during init not handled by the kernel?
Shouldn't they?
How can a driver return an error during init without causing kernel panics?

pabigot reviewed (changes requested):
The way the init function blocks isn't acceptable.

joelrietzle commented:
If I want to use the sample together with a nrf9160dk and combine it with the lwm2m_client. Would that be possible to do in the same way that the accelerometer is implemented for example?

bbilas commented:
@GeorgeGkinis are you gonna resolve all conflicts and prepare this series to be ready to merge? That would be nice to see it in the mainline.

joelrietzle commented:
I would LOVE to see this in production as well!!

ons 12 maj 2021 kl. 16:44 skrev Bartosz Biłas ***@***.***>:

> @GeorgeGkinis <https://github.com/GeorgeGkinis> are you gonna resolve all
> conflicts and prepare this series to be ready to merge? That would be nice
> to see it in the mainline.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/24295#issuecomment-839830485>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQGMRUABOAYN4JA4DI6OJOTTNKH3NANCNFSM4MGMJBAA>
> .
>

nobodyguy commented:
I'm working on porting this driver to upstream, but I'd rather create a new PR.

GeorgeGkinis commented:
> I'm working on porting this driver to upstream, but I'd rather create a new PR.

Could you share your PR please?

nobodyguy commented:
> Could you share your PR please?
Sorry for the delay, I had to postpone this because of update of my project to the upstream Zephyr + various HW issues.
But this is currently a top priority task for me and I've been working on this since yesterday, so I'll share a PR link once it's finished.

nobodyguy commented:
@GeorgeGkinis Just to give you an update after 1 week of occasional work:
I ported the code to the upstream Zephyr, implemented all the feedback (except init blocking) from this PR, added PM subsystem support and basically got stuck on the initialization issue.

Regarding to the blocking driver's init for up to 600ms:
The datasheet states: "_Settling time (400ms) refers to the time from power up, reset, input channel change and gain change
to valid stable output data._"
Does it mean it takes this long for data to be ready (transition of data pin from high to low)? Or is the data ready earlier but just noisy and unusable?
Because there are different options how to handle these situations.

GeorgeGkinis commented:
Hey there @nobodyguy,

I believe it means that the measurement cannot be trusted.
Also there is no PULL DOWN on the SCK pin until HX711 is ready (~400ms after power on)  

To be honest I think that the driver should work on a interrupt driven manner.

I have no idea how to init() without actually taking a measurement and potentially block until the first PULL DOWN on the sck_gpio pin. Maybe just do everything init() does without hx711_sample_fetch(). Then setup an interrupt for every time when SCK goes LOW and take a measurement, which also sets up GAIN.

After that everytime hx711_sample_fetch() is invoked the driver just returns the last measurement.
If more than 400ms have elapsed without SCK going LOW then return a ERROR code.

Problem is I have issues setting up my dev environment on windows, and after that I need to study the interrupt mechanisms of Zephyr.

I purchased an extra PC to be able to work under linux and am setting up my DEV environment again.

I hate to let al those hours already spend go to waste :/
I would like to finish this or assist you if you are more adept than me. You can also contact me at **gkinis.georgios@gmail.com**

GeorgeGkinis commented:
@carlescufi How do I reopen this PR?

MaureenHelm commented:
> @carlescufi How do I reopen this PR?

GitHub won't let you reopen a PR if you force push the branch after the PR was closed. Please create a new PR and link back to this one in the description.

GeorgeGkinis commented:
Thanks for the explanation.

We are working with @nobodyguy and will make a new PR when done.
Thanks!

GeorgeGkinis commented:
All review are resolved except for [this one ](https://github.com/zephyrproject-rtos/zephyr/pull/24295#pullrequestreview-590524238).

Created a [new branch](https://github.com/GeorgeGkinis/zephyr/tree/hx711-driver).

I will be creating a new PR shortly.

GeorgeGkinis commented:
I have created a new pull request: [#42570](https://github.com/zephyrproject-rtos/zephyr/pull/42570)

Thanks.
